src/dataMap/metaManager/lazyFactoryMap.js

```diff
@@ -0,0 +1,265 @@
+ *               ├─────────────┐
+ *              \│/            │
+ * +------+------+------+------+------+
+ * | 0/10 | 1/11 | 2/12 | 3/13 | 4/14 |  Keys before
+ * +------+------+------+------+------+
+ *
+ * +------+------+------+
+ * | 0/10 | 1/11 | 2/14 |  Keys after
+ * +------+------+------+
+ *    │       │      │
+ *    │       │      └─────────────┐
+ *    │       │                    │
+ *    │       │                    │
+ *    │       │                    │
+ *    │       │                    │
+ *    │       │                    │
+ * +------+------+------+------+------+
+ * | AAA  | BBB  | CCC  | DDD  | EEE  |  Data
+ * +------+------+------+------+------+
+ *                  /│\   /|\
+ *                   └──┬──┘
+ *         This data is marked as "hole" which
+ *         means that can be replaced by new item
+ *         when that will be created.
+ *
+ * map.obtain(2) // returns the old value ("EEE") as it should
+ */
+export default class LazyFactoryMap {
+  constructor(valueFactory) {
+    this.valueFactory = valueFactory;
+    /**
+     * An array which contains data.
+     *
+     * @type {Array}
+     */
+    this.data = [];
+    /**
+     * An array of indexes where the key of the array is mapped to the value which points to the
+     * specific position of the data array.
+     *
+     * @type {Number[]}
+     */
+    this.index = [];
+    /**
+     * An array of indexes where points to the data items which can be replaced by obtaining new
+     * ones. The "holes" are a side effect of deleting entries.
+     *
+     * @type {Number[]}
+     */
+    this.holes = [];
+  }
+
+  /**
+   * Gets or if data not exist creates and returns new data.
+   *
+   * @param {Number} key The item key as zero-based index.
+   * @returns {*}
+   */
+  obtain(key) {
+    assert(() => isFiniteUnsignedNumber(key), 'Expecting an unsigned finite number.');
+
+    const dataIndex = this._getStorageIndexByKey(key);
+    let result;
+
+    if (dataIndex >= 0) {
+      result = this.data[dataIndex];
+
+      if (result === void 0) {
+        result = this.valueFactory(key);
+        this.data[dataIndex] = result;
+      }
+    } else {
+      result = this.valueFactory(key);
+
+      if (this.holes.length > 0) {
+        const reuseIndex = this.holes.pop();
+
+        this.data[reuseIndex] = result;
+        this.index[key] = reuseIndex;
+      } else {
+        this.data.push(result);
+        this.index[key] = this.data.length - 1;
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * Inserts an empty data to the map. This method creates an empty space for obtaining
+   * new data.
+   *
+   * @param {Number} key The key as volatile zero-based index at which to begin inserting space for new data.
+   * @param {Number} [amount=1] Ammount data to insert.
+   */
+  insert(key, amount = 1) {
+    assert(() => (isFiniteUnsignedNumber(key) || isNullish(key)), 'Expecting an unsigned finite number or null/undefined argument.');
```

Comment:
Maybe `isNumber` would be sufficient? Maybe this `isNullish` check should be removed and this should be checked outside the module?

Reply:
I'll check if the `Number.isInteger` the function is sufficient for this task 👍 The nullish check has to stay because that method currently accepts Numbers, `null` or `undefined`.

Reply:
Fixed within https://github.com/handsontable/handsontable/pull/6254/commits/eb1f3258945f22e65a1255460a62cc3da680e58e#diff-7cffc9b2d87e9b8c5d2eec7ae2cab989R60